lib/Models/Catalog/Esri/ArcGisImageServerCatalogItem.ts

```diff
@@ -0,0 +1,614 @@
+    const labelsRegex = /_Labels$/;
+
+    let items: StratumFromTraits<LegendItemTraits>[] = [];
+
+    (this._legends?.layers || []).forEach((l) => {
+      if (noDataRegex.test(l.layerName) || labelsRegex.test(l.layerName)) {
+        return;
+      }
+
+      l.legend?.forEach((leg) => {
+        const title = replaceUnderscores(
+          leg.label !== "" ? leg.label : l.layerName
+        );
+        const dataUrl = "data:" + leg.contentType + ";base64," + leg.imageData;
+        items.push(
+          createStratumInstance(LegendItemTraits, {
+            title,
+            imageUrl: dataUrl,
+            imageWidth: leg.width,
+            imageHeight: leg.height
+          })
+        );
+      });
+    });
+
+    items = uniqWith(items, (a, b) => a.imageUrl === b.imageUrl);
+
+    return [createStratumInstance(LegendTraits, { items })];
+  }
+}
+
+StratumOrder.addLoadStratum(ImageServerLegendStratum.stratumName);
+
+export default class ArcGisImageServerCatalogItem extends UrlMixin(
+  DiscretelyTimeVaryingMixin(
+    MinMaxLevelMixin(
+      CatalogMemberMixin(
+        MappableMixin(CreateModel(ArcGisImageServerCatalogItemTraits))
+      )
+    )
+  )
+) {
+  static readonly type = "esri-imageServer";
+
+  private _legendStratumLoader = new AsyncLoader(
+    this.forceLoadLegends.bind(this)
+  );
+
+  constructor(...args: ModelConstructorParameters) {
+    super(...args);
+    makeObservable(this);
+  }
+
+  get typeName() {
+    return i18next.t("models.arcGisImageServerCatalogItem.name");
+  }
+
+  get type() {
+    return ArcGisImageServerCatalogItem.type;
+  }
+
+  protected async forceLoadMetadata(): Promise<void> {
+    const stratum = await ImageServerStratum.load(this);
+    runInAction(() => {
+      this.strata.set(ImageServerStratum.stratumName, stratum);
+    });
+    await this._legendStratumLoader.load();
+  }
+
+  protected async forceLoadLegends(): Promise<void> {
+    const stratum = await ImageServerLegendStratum.load(this);
+    runInAction(() => {
+      this.strata.set(ImageServerLegendStratum.stratumName, stratum);
+    });
+  }
+
+  protected forceLoadMapItems(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  @computed
+  get discreteTimes() {
+    const imageServerStratum: ImageServerStratum | undefined = this.strata.get(
+      ImageServerStratum.stratumName
+    ) as ImageServerStratum | undefined;
+
+    if (imageServerStratum?.imageServer.timeInfo === undefined)
+      return undefined;
+
+    const result: (StratumFromTraits<DiscreteTimeTraits> & {
+      time: string;
+    })[] = [];
+
+    createDiscreteTimesFromIsoSegments(
+      result,
+      new Date(
+        imageServerStratum.imageServer.timeInfo.timeExtent[0]
+      ).toISOString(),
+      new Date(
+        imageServerStratum.imageServer.timeInfo.timeExtent[1]
+      ).toISOString(),
+      undefined,
+      this.maxRefreshIntervals
+    );
+    return result;
+  }
+
+  private get _currentImageryParts(): ImageryParts | undefined {
```

Comment:
Nitpick: might be better to also make this `@computed`

Reply:
Done!